archive_extractor/limiting_aggregating_read_closer.go

```diff
@@ -0,0 +1,53 @@
+package archive_extractor
+
+import (
+	"errors"
+	"io"
+)
+
+var ErrTooManyEntries = errors.New("too many entries in archive")
+var ErrCompressLimitReached = errors.New("total bytes limit reached")
+
+type LimitAggregatingReadCloserProvider struct {
+	Total int64
+	Limit int64
+}
+
+func (provider *LimitAggregatingReadCloserProvider) CreateLimitAggregatingReadCloser(rc io.Reader) LimitAggregatingReadCloser {
+	return &limitAggregatingReadCloser{
+		Reader: rc,
+		Total:  &provider.Total,
+		Limit:  provider.Limit,
+	}
+}
+
+type LimitAggregatingReadCloser interface {
+	Read(p []byte) (int, error)
+	Close() error
+}
+
+type limitAggregatingReadCloser struct {
+	Reader io.Reader
+	Total  *int64
```

Comment:
why pointer?

Reply:
For the aggregation.
Every time a limitAggregatingReadCloser increase the total, I would like the provider to see this increase. All following limitAggregatingReadCloser will see the same total.
This will cause all limitAggregatingReadCloser to count against the same total, so if you have one archive with multiple entries, multiple limitAggregatingReadClosers will be created, one per entry, but each time the total is increased, the increment will be saved to the provider and following created limitAggregatingReadClosers will start with an higher total.